#16143: Inplace support for binary_ng ops with fused activations

mcw-anasuya opened the pull request:
### Ticket
#16143
#16153 

### Problem description
Provide inplace support for binary_ng ops with fused activations with validation checks for input.

### What's changed
Added in-place support for 20 ops 
Added fused activations on input tensors.
Added validation checks for input tensor a.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12966520490 - PASSED
- [x] Blackhole post-commit tests https://github.com/tenstorrent/tt-metal/actions/runs/12966965482 - PASSED
- [x] (Single-card) Tests for new models https://github.com/tenstorrent/tt-metal/actions/runs/12966971183 - same as main
- [x] New/Existing tests provide coverage for changes

Commit:
#16153: Implement binary-ng fused input activations

mcw-anasuya committed:
#16143: Inplace support for binary_ng ops with fused activations

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +105,81 @@ Tensor BinaryNg<binary_op_type>::invoke(
+    tt::stl::Span<const ttnn::operations::unary::UnaryOpType> lhs_activations,
+    tt::stl::Span<const ttnn::operations::unary::UnaryOpType> rhs_activations,
+    tt::stl::Span<const ttnn::operations::unary::UnaryOpType> post_activations) {
+    auto input_a = typecast_to(DataType::BFLOAT16, input_tensor_a);
```
are we typecasting all inputs to bf16 ?

KalaivaniMCW replied:
are we using input_a and input_b ?

umadevimcw commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +105,81 @@ Tensor BinaryNg<binary_op_type>::invoke(
+
+    return ttnn::prim::binary_ng(
+        queue_id,
+        input_tensor_a,
```
`input_tensor_a` and `input_tensor_b` should be replaced by `input_a` and `input_b`

umadevimcw commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +105,81 @@ Tensor BinaryNg<binary_op_type>::invoke(
+    return ttnn::prim::binary_ng(
+        queue_id,
+        input_tensor_a,
+        scalar,
```
input_tensor_a  should be replaced by input_a

umadevimcw replied:
Please ensure the `test_bcast_tcast.py` file is passed after this change.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp:
```diff
@@ -163,6 +164,16 @@ BinaryNgDeviceOperation::program_factory_t BinaryNgDeviceOperation::select_progr
+
     if (output_tensor.has_value()) {
         return output_tensor.value();
     }
```
can we have a single if check for output_tensor

```
   if (output_tensor.has_value()) {
         if (input_tensor_a.buffer()->address() == output_tensor->buffer()->address()) {
             const auto& output_shape = compute_output_specs(operation_attributes, tensor_args);
             TT_FATAL(
                 input_tensor_a.logical_shape() == output_shape.logical_shape(),
                 "In-place operation rule violation: Input tensor does not support output tensor shape.");
         }
         
              return output_tensor.value();
     }
```

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp:
```diff
@@ -102,6 +102,7 @@ void BinaryNgDeviceOperation::validate_on_program_cache_hit(
     const int rank_a = input_shape_a.rank();
     const int rank_b = input_shape_b.rank();
     const int larger_rank = std::max(rank_a, rank_b);
+
```
remove

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +105,106 @@ Tensor BinaryNg<binary_op_type>::invoke(
+        binary_op_type,
+        output_dtype,
+        input_tensor_a.memory_config(),
+        input_tensor_a,
```
The buffer check will not match as `input_a` and `input_tensor_a` are different
need to pass `input_a` as optional_output here, then copy the result back to `input_tensor_a`. This is for only when typecast needs to happen

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,158 @@ Tensor BinaryNg<binary_op_type>::invoke(
+            rhs_activations,
+            post_activations);
+    } else {
+        auto input_a = typecast_to(DataType::BFLOAT16, input_tensor_a);
```
input_b would also need typecast to bf16

mcw-anasuya replied:
input_b is a scalar. Is it required to typecast scalar as well?

mcw-anasuya replied:
The scalar is being converted to bf16 in binary_ng_program_factory.cpp

umadevimcw commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,158 @@ Tensor BinaryNg<binary_op_type>::invoke(
+            post_activations);
+    } else {
+        auto input_a = needs_typecast_a ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;
+        auto input_b = needs_typecast_b ? typecast_to(DataType::BFLOAT16, input_tensor_b) : input_tensor_b;
```
Need to simplify this/ modularise it and re use it

umadevimcw commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,158 @@ Tensor BinaryNg<binary_op_type>::invoke(
+            lhs_activations,
+            rhs_activations,
+            post_activations);
+    } else {
```
same here as well

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,131 @@ Tensor BinaryNg<binary_op_type>::invoke(
+            rhs_activations,
+            post_activations);
+    } else {
+        Tensor input_a = check_dtype(input_tensor_a) ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;
```
Tensor input_a = typecast_a ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,131 @@ Tensor BinaryNg<binary_op_type>::invoke(
+            post_activations);
+    } else {
+        Tensor input_a = check_dtype(input_tensor_a) ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;
+        Tensor input_b = check_dtype(input_tensor_b) ? typecast_to(DataType::BFLOAT16, input_tensor_b) : input_tensor_a;
```
Tensor input_b = typecast_b ? typecast_to(DataType::BFLOAT16, input_tensor_b) : input_tensor_b;

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,131 @@ Tensor BinaryNg<binary_op_type>::invoke(
+            rhs_activations,
+            post_activations);
+    } else {
+        Tensor input_a = check_dtype(input_tensor_a) ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;
```
Tensor input_a = typecast_a ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp:
```diff
@@ -111,6 +111,16 @@ void BinaryNgDeviceOperation::validate_on_program_cache_hit(
             i,
             a_dim,
             b_dim);
+        if (input_tensor_a.buffer()->address() == output_tensor->buffer()->address()) {
+            TT_FATAL(rank_a >= rank_b, "In-place operation rule violation for rank a: {}, rank b: {}", rank_a, rank_b);
+
+            TT_FATAL(
+                a_dim >= b_dim,
+                "In-place operation rule violation for dimension {}, dim a: {}, dim b: {}",
+                i,
+                a_dim,
+                b_dim);
+        }
```
Do we have tests for these assertions? Also the assertion for `rank_a >= rank_b` can be moved outside the loop.

mcw-anasuya replied:
These assertions are tested in test_inplace_binary_scalar_ops_invalid_bcast.
Sure, I'll move the rank assertion outside the loop.

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.hpp:
```diff
@@ -57,12 +57,49 @@ struct BinaryNg {
+
 }  // namespace ttnn::operations::binary_ng
 
 inline Tensor typecast_to(DataType dtype, const Tensor& input) {
     return input.get_dtype() == dtype ? input : ttnn::typecast(input, dtype);
 }
 
+inline bool check_dtype(const Tensor& input) {
+    return (input.get_dtype() == DataType::BFLOAT8_B || input.get_dtype() == DataType::BFLOAT4_B);
+}
+
```
Neither of these functions needs to be in the `hpp` file, can we move them to the `cpp`? Also `check_dtype` is not a helpful name, maybe `needs_typecast_to_bfloat16`?

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +106,131 @@ Tensor BinaryNg<binary_op_type>::invoke(
+    tt::stl::Span<const ttnn::operations::unary::UnaryOpType> lhs_activations,
+    tt::stl::Span<const ttnn::operations::unary::UnaryOpType> rhs_activations,
+    tt::stl::Span<const ttnn::operations::unary::UnaryOpType> post_activations) {
+    bool typecast_a = check_dtype(input_tensor_a);
+    bool typecast_b = check_dtype(input_tensor_b);
+
+    if (!typecast_a && !typecast_b) {
+        return ttnn::prim::binary_ng(
+            queue_id,
+            input_tensor_a,
+            input_tensor_b,
+            binary_op_type,
+            input_tensor_a.get_dtype(),
+            input_tensor_a.memory_config(),
+            input_tensor_a,
+            lhs_activations,
+            rhs_activations,
+            post_activations);
+    } else {
+        Tensor input_a = typecast_a ? typecast_to(DataType::BFLOAT16, input_tensor_a) : input_tensor_a;
+        Tensor input_b = typecast_b ? typecast_to(DataType::BFLOAT16, input_tensor_b) : input_tensor_b;
```
Why is the condition for typecasting different between the existing and inplace operations?

mcw-anasuya replied:
The input is only being typecasted if it is in bf8b or bf4b format.
We need to pass the input as it is for bf16, float32 and in some cases int32 formats without typecasting it. Fp32 and bitwise ops using int32 formats are supported in this PR https://github.com/tenstorrent/tt-metal/pull/16523

patrickroberts replied:
The reason I was asking was because I wanted to know whether we should update the typecasting in the existing operations to have the same conditions as the inplace operations you have here, then we can implement these in terms of the existing operations by passing `input_tensor_a` as the `optional_output_tensor` and avoid duplicating the logic for typecasting in two different places.

patrickroberts replied:
Other than this one suggestion, everything else looks good to me

KalaivaniMCW replied:
@patrickroberts Yes, I think we should update the typecasting in the existing operations i.e. when an optional_output_tensor is passed by the user then we need to perform these shape & dtype checks there as well and then typecast the result from bf16 to expected dtype of optional_output_tensor.

patrickroberts replied:
Sounds good, I'll approve this PR as-is, but I created a new issue so we can refactor this and unify the logic between them.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp:
```diff
@@ -194,6 +198,15 @@ void BinaryNgDeviceOperation::validate_on_program_cache_hit(
+        if (output_tensor.has_value() && (input_tensor_a.buffer()->address() == output_tensor->buffer()->address())) {
+            TT_FATAL(
+                a_dim >= b_dim,
+                "In-place operation rule violation for dimension {}, dim a: {}, dim b: {}",
```
Please clarify the rule for a user. Assume that  user can’t go to the source code to read the above if statement. 

https://github.com/tenstorrent/tt-metal/blob/main/contributing/ErrorMessageBestPractices.md

mcw-anasuya replied:
@ayerofieiev-tt I removed this check since the shape mismatch is being handled in #16523 which is merged to main. I modified the test for invalid cases accordingly.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp:
```diff
@@ -176,6 +176,10 @@ void BinaryNgDeviceOperation::validate_on_program_cache_hit(
     const int rank_b = input_shape_b.rank();
     const int larger_rank = std::max(rank_a, rank_b);
+    if (output_tensor.has_value() && (input_tensor_a.buffer()->address() == output_tensor->buffer()->address())) {
+        TT_FATAL(rank_a >= rank_b, "In-place operation rule violation for rank a: {}, rank b: {}", rank_a, rank_b);
```
Please see my other comment below

mcw-anasuya replied:
Same here as well 
https://github.com/tenstorrent/tt-metal/pull/16449#discussion_r1929571049

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/binary_ng.cpp:
```diff
@@ -105,6 +114,132 @@ Tensor BinaryNg<binary_op_type>::invoke(
+    bool typecast_a = needs_typecast_to_bfloat16(input_tensor_a);
+    bool typecast_b = needs_typecast_to_bfloat16(input_tensor_b);
+
+    if (!typecast_a && !typecast_b) {
```
To double check, is it intentional that we only support or a) nothing needs typecasting or b) both need?

mcw-anasuya replied:
@ayerofieiev-tt Yes, if both tensors do not require typecasting, we pass the input tensors as they are and avoid creating two new tensors. We only create the two tensors if any one of them or both require typecasting (if typecasting is not required on `input_tensor_a` but is required on `input_tensor_b`, then `input_a` will contain the original `input_tensor_a` while `input_b` will contain the typecasted `input_tensor_b`.

patrickroberts commented:
I've fixed the merge conflicts that are currently here and I'm testing that the build compiles and the existing tests pass before I force push. Just wanted to share this here so that nobody duplicates efforts on this since it's already done. @mcw-anasuya @umadevimcw

patrickroberts commented:
> I've fixed the merge conflicts that are currently here and I'm testing that the build compiles and the existing tests pass before I force push. Just wanted to share this here so that nobody duplicates efforts on this since it's already done. @mcw-anasuya @umadevimcw

Turns out I have the commit order messed up so I would have to redo the rebase since it makes me fix the exact same conflicts I already fixed just to reorder the commits from what I have now, which is here: https://github.com/tenstorrent/tt-metal/tree/proberts/inplace_binary_ng
This shows what the code should look like for this PR but it's not really usable for fixing the merge conflicts that are currently present unfortunately because of what I explained above.

mcw-anasuya commented:
> > I've fixed the merge conflicts that are currently here and I'm testing that the build compiles and the existing tests pass before I force push. Just wanted to share this here so that nobody duplicates efforts on this since it's already done. @mcw-anasuya @umadevimcw
> 
> Turns out I have the commit order messed up so I would have to redo the rebase since it makes me fix the exact same conflicts I already fixed just to reorder the commits from what I have now, which is here: https://github.com/tenstorrent/tt-metal/tree/proberts/inplace_binary_ng This shows what the code should look like for this PR but it's not really usable for fixing the merge conflicts that are currently present unfortunately because of what I explained above.

@patrickroberts I have rebased the branch to main and pushed the changes done here https://github.com/tenstorrent/tt-metal/commit/5be36dc7ddcf703403af4ccb99a7d28833225d24
